### ISSUE TITLE
Pass cfg to build script

### ISSUE THREAD

**Issue by dtolnay** [rust-lang/docs.rs#1580]
https://github.com/rust-lang/docs.rs/pull/1543 broke proc-macro2's documentation build &mdash; see https://github.com/dtolnay/proc-macro2/issues/309.

As far as I can tell the difference is that `cargo rustdoc -- --cfg mycfg` sets mycfg during **both** the compilation of the build.rs and the invocation of rustdoc, whereas `cargo rustdoc --config 'build.rustdocflags=["--cfg", "mycfg"]'` sets it only during rustdoc and **not** the compilation of build.rs.

As far as I can tell there is now no way to set cfg for a build script build on docs.rs.

@willcrichton @jyn514

**Comment by willcrichton** [comment]
Thanks for the report @dtolnay! 

@jyn514 what do you think is the right fix here? We could add `cfg` as a new field to `[package.metadata.docs.rs]` if it's a special case. Or I could duplicate the `rustdoc-args` field between `--config 'build.rustdocflags=<flags>'` and `rustdoc -- <flags>`. Not sure if that would also cause duplicate flag CLI errors.

**Comment by jyn514** [comment]
> As far as I can tell there is now no way to set cfg for a build script build on docs.rs.

Can't you set it with `RUSTFLAGS`? There's a lever for modifying it in the metadata with `rustc-args`.

Hmm, I guess that applies it to all dependencies, too ... is that a problem for your use-case?

**Comment by dtolnay** [comment]
Nope, cfg from `rustc-args` does not get applied either. That was already in the build that regressed. See https://github.com/dtolnay/proc-macro2/blob/1.0.33/Cargo.toml#L19-L20 and https://docs.rs/crate/proc-macro2/1.0.33/builds/473000.

**Comment by Nemo157** [comment]
> As far as I can tell the difference is that `cargo rustdoc -- --cfg mycfg` sets mycfg during **both** the compilation of the build.rs and the invocation of rustdoc

This seems like it would be incorrect behavior, and doesn't match what I see testing it locally. I also tried building the `proc-macro2==1.0.33` docs with my script that behaves similarly to how docs.rs used to, which failed similarly to that build. That ended up running the command:

```console
env DOCS_RS=1 RUSTCFLAGS='--cfg procmacro2_semver_exempt' cargo rustdoc --locked --target x86_64-unknown-linux-gnu -- -Z unstable-options --cfg procmacro2_semver_exempt --cfg doc_cfg --extern-html-root-url quote=https://docs.rs/quote/1.0.10 --extern-html-root-url unicode_xid=https://docs.rs/unicode-xid/0.2.2
```

**Comment by jyn514** [comment]
> RUSTCFLAGS='--cfg procmacro2_semver_exempt' 

Shouldn't that be RUSTFLAGS?

**Comment by Nemo157** [comment]
Yep, but using `RUSTFLAGS` doesn't fix it, the `--cfg` doesn't appear to get passed to the build script build.

**Comment by Nemo157** [comment]
I'm really confused, copying the exact command run from the [previous successful build](https://docs.rs/crate/proc-macro2/1.0.32/builds/454089) while trying to build `proc-macro2==1.0.32` locally also fails in the same way, even when overriding to the nightly that was used.

```
"cargo" "+nightly-2021-10-26" "rustdoc" "--lib" "-Zrustdoc-map" "-Z" "unstable-options" "--config" "build.rustflags=[\"--cfg\", \"procmacro2_semver_exempt\"]" "-Zunstable-options" "--config=doc.extern-map.registries.crates-io=\"https://docs.rs/{pkg_name}/{version}/x86_64-unknown-linux-gnu\"" "-j2" "--" "--cfg" "procmacro2_semver_exempt" "--cfg" "doc_cfg" "-Z" "unstable-options" "--emit=invocation-specific" "--resource-suffix" "-20211026-1.58.0-nightly-e269e6bf4" "--static-root-path" "/" "--cap-lints" "warn" "--disable-per-crate-search"
```

**Comment by Nemo157** [comment]
Ok, found the reason I was having that fail too, `--config build.rustflags` gets overridden by `target.x86_64-unknown-linux-gnu.rustflags` in my global config instead of being merged together. That shouldn't affect docs.rs since it shouldn't have a global `.cargo/config`.

After trying a bunch of changes I finally found the difference, the `--target x86_64-unknown-linux-gnu` changes whether `build.rustflags` gets passed to the build-script or not. That seems like a cargo bug to me.

**Comment by jyn514** [comment]
Ah, ok - that's caused by https://github.com/rust-lang/docs.rs/pull/1559 then, not https://github.com/rust-lang/docs.rs/pull/1543.

**Comment by jyn514** [comment]
cc https://github.com/rust-lang/cargo/issues/7677

**Comment by Nemo157** [comment]
And for the specific failure here: https://github.com/rust-lang/cargo/issues/4423. I've always assumed `build.rustflags` was global arguments to apply to all builds, but it looks like it actually means "the args for the implicit host target, but only for host-only builds like proc-macros and build-scripts when the final artifact is the implicit host target"; and there's no way to explicitly set flags for "host-only non-artifact builds".

**Comment by Nemo157** [comment]
Ah, following links a little more I found https://doc.rust-lang.org/nightly/cargo/reference/unstable.html#host-config which might be able to do what we need; `-Zhost-config --config=host.rustflags=...`.

**Comment by Nemo157** [comment]
To clarify what happened and how I think we should fix this:

Prior to #1559:
* compiling for `x86_64-unknown-linux-gnu` would apply the specified `rustc-args` to all builds, including while compiling the build-scripts
* compiling for other targets would apply the specified `rustc-args` to only the builds that get linked into the final artifact, so not including the build-scripts

After #1559:
* all (non-proc-macro) builds are now treated as cross-compiles
* compiling for all targets would apply the specified `rustc-args` to only the artifact builds, not including the build-scripts

I think we should update to apply the specified `rustc-args` to all builds no matter what the `--target` is. I attempted to use `-Zhost-config` to implement this, but ran into <https://github.com/rust-lang/cargo/issues/10206>; so I think we're blocked on fixing this till that is fixed in cargo.

**Comment by jyn514** [comment]
https://github.com/rust-lang/cargo/issues/10206 has been fixed.
